x/sequencer/keeper/rotation.go

```diff
@@ -0,0 +1,133 @@
+package keeper
+
+import (
+	"sort"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/dymensionxyz/dymension/v3/x/sequencer/types"
+)
+
+func (k Keeper) startNoticePeriodForSequencer(ctx sdk.Context, seq *types.Sequencer) time.Time {
+	completionTime := ctx.BlockTime().Add(k.NoticePeriod(ctx))
+	seq.NoticePeriodTime = completionTime
+
+	k.UpdateSequencer(ctx, *seq)
+	k.AddSequencerToNoticePeriodQueue(ctx, *seq)
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeNoticePeriodStarted,
+			sdk.NewAttribute(types.AttributeKeyRollappId, seq.RollappId),
+			sdk.NewAttribute(types.AttributeKeySequencer, seq.Address),
+			sdk.NewAttribute(types.AttributeKeyCompletionTime, completionTime.String()),
+		),
+	)
+
+	return completionTime
+}
+
+// MatureSequencersWithNoticePeriod start rotation flow for all sequencers that have finished their notice period
+// The next proposer is set to the next bonded sequencer
+// The hub will expect a "last state update" from the sequencer to start unbonding
+func (k Keeper) MatureSequencersWithNoticePeriod(ctx sdk.Context, currTime time.Time) {
+	seqs := k.GetMatureNoticePeriodSequencers(ctx, currTime)
+	for _, seq := range seqs {
+		if k.isProposer(ctx, seq.RollappId, seq.Address) {
```

Comment:
can you document this wrt. non proposer who is in notice period

as I understand, non proposer in notice period must be next proposer
so when rotation completes he become proposer and will immediately finish notice period here

So notice period not working as intended, as a notice from the time they become proposer..